README.md
# Philter v1.5.0
[![npm](https://img.shields.io/npm/v/philter.svg)](https://www.npmjs.com/package/philter) [![dependencies](https://david-dm.org/specro/philter.svg)](https://david-dm.org/specro/philter)

Philter is a JS plugin giving you the power to control CSS filters with HTML data attributes.
Visit the [Demo page](http://specro.github.io/Philter/) for examples.

## Installation
Since version 1.3.0 Philter comes as vanilla js plugin or npm package.
Download the plugin and move the `philter` directory to your `js` directory, then include it in your page:
```html
<script src="js/philter/philter.min.js"></script>
```
or with NPM:
```shell
npm install philter
```
## Usage
### Node
```js
const philter = require('philter')

philter(['index.html', 'post.html'], { tag: true, customFilterDir: '', customFilters: [] } (css, svg) => {
  console.log('CSS: ', css)
  console.log('SVG: ', svg)
})
```
You can also pass 3 parameters to philter:
* tag - boolean - This enables the 'philter' part in data-philter-<filter>. If you don't use any plugins which use data attributes or they won't collide with Philter, you can set this to false to omit this part and shorten your markup.
* customFilterDir - string - Directory where custom filters are stored.
* customFilters - array - Array of custom filter names.

### Browser
I highly recommend using Philter in Node and use the plugin version only for development and demonstration purposes, since the browser version doesn't support all the listed filters. 

First initiate the plugin:
```html
<script>
  new Philter({
    transitionTime: 0.5, // hover transition time
    tag: true // 'philter' in data attributes
  });
</script>
```
You can pass 3 parameters to Philter:
* transitionTime - The hover transition time of default CSS filters
* tag - This enables the 'philter' part in data-philter-<filter>. If you don't use any plugins which use data attributes or they won't collide with Philter, you can set this to false to omit this part and shorten your markup.

### CLI
```shell
philter index.html post.html -c index.html -s index.html
```

```
Usage: philter [options] <file ...>

Options:

  -h, --help                     output usage information
  -V, --version                  output the version number
  -n, --no-tag                   No "philter" in data attributes
  -s, --svg <dir>                SVG directory or svg/html file to append to
  -c, --css <dir>                CSS directory or css/html file to append to
  -H, --html                     Pass HTML instead of filenames
  -D, --custom-filter-dir <dir>  Custom SVG filter directory
  -F, --custom-filters <list>    Custom filters

```

## Format

Now you can start using the filters. The plugin uses this kind of syntax format:
```html
data-philter-<filter>="<value>"
```
or
```html
data-philter-<filter>="<value> <hover-value>"
```
You give an element the data attribute for a specific filter and then a value for it. You can also add another value that the filter will use when hovering on that element.
For example:
```html
<div data-philter-blur="5"></div>
```
This element would be blured in 5px radius. If we would add another value, like this:
```html
<div data-philter-blur="5 0"></div>
```
The element would unblur when hovered over with the mouse.
With filters that use more than one value you have to specify every value for hover too.
You can add more than one filter onto an element by using the same method:
```html
<div data-philter-blur="5 0" data-philter-grayscale="100"></div>
```
Philter even supports custom SVG filters:
```html
<div data-philter-blur="5 0" data-philter-svg="filter"></div>
```
Where 'filter' in 'data-philter-svg' attribute is the ID of the filter.
Also Philter has pre-built custom filters:
```html
<div data-philter-color="#00ff00 50"></div>
```
This one would overlay the element with #00ff00 color in 50% opacity.
More filters are to come in the near future. You have any suggestions or know a filter that certainly has to be present in Philter? Just contact me or Elephento team.

## More info on filters
Here's a list of filters that you can use and their limitations in Philter.
* blur
* grayscale
* hue-rotate
* saturate
* sepia
* contrast
* invert
* opacity
* brightness
* drop-shadow - Requires 4 values. In the browser the 4th value instead of color is opacity 0 to 100%, color is locked to black.
* svg - Custom SVG filter. Requires 1 value - filter ID.
* color - Requires 2 values. Color and opacity.
* vintage - Requires an integer from 1 to 6.
* duotone - Requires 2 values. 2 colors in hex.
* custom - Requires a string - custom filter name.
* anaglyph - Experimental - Requires an anaglyph offset value.

Drop shadow filter in browser supports only black color because with it's already long class it would be even longer with rgba implementation.

### Vintage
There are 6 vintage filters:
* Rises contrast. Brings out details and colors.
* Washes out the image with light brown sepia.
* Raises the brightness and gives a green/cyan look.
* Close to 3 but a bit less brightness and more green.
* Close to 2 but mixed with violet. Gives a sweet/daydream look.
* Grayscale but better (IMO :))

### Duotone
Duotone filter maps shadows and highlights of the image to 2 different colors. First color is mapped to the shadow and second to the highlights. Colors for this filter should be provided in hex format e.g., #123456.

### Experimental
This section contains filters that are experimental i.e. work only in one browser, have image breaking bugs (even though this whole SVG thing is pretty buggy) etc. These filters will be showcased here: [Philter experimental](https://specro.github.io/Philter/experimental).

Anaglyph filter tries to imitate the [3D anaglyph effect](https://en.wikipedia.org/wiki/Anaglyph_3D) that is widely used for 3D images and graphic effects. So far this should work 100% only on Chrome. Also I use feBlend to cut off the offset part which means that the bigger the offset is the more of the image is being cut off the left side.

### Custom
You can use filters that you wrote by yourself in NPM/CLI version by using the custom tag like this:
```html
data-philter-cutom="<custom-filter-name>"
```
If using custom filters you must supply the directory where they're stored and custom filter names in the options. The file of the filter must have that name and its id must be that same name.

## Compatibility
Philter was developed and tested on Chrome 46+, Firefox 41+, Opera 34+ and Edge 20+. The default CSS filters should be compatible with most versions of browsers that support filters. The custom filters are supported only by Firefox, Chrome and Opera. You may notice glitching on Edge when more than one hover element is on the page and loss of some filters when they are stacked on one element.

## Issues
This is mainly due to SVG filter limitations or complexities. It may be solved in the future... or it may not.
* On my recent tests with Chrome SVG filters stack with other filters but as always you may encounter bugs.
* SVG filters don't support CSS transitions.
* SVG filters actually know what to do on hover but ^ and you may encounter other bugs (like flickering and so on). Especially on Edge and IE browsers.

## WIP
I'm working on all sorts of stuff that involves this plugin and doesn't. So please bear with the way I develop Philter. If you have any suggestions ideas or just wanna say something you can send me an email at liudas.dzisevicius@gmail.com or tweet @baldassertation.
* Gulp (I work with Gulp, so there will be no Grunt here. Sorry.)
* Webpack
* More custom SVG filters

## License
Philter is licensed under MIT License.


CHANGELOG.md
# Change Log
## [Unreleased]

## [1.5.0] - 2018-06-03
### Added
- Duotone filter

### Changed
- Moved SVG files in the browser plugin into JS to remove unneeded requests

### Fixed
- Custom filter paths resolving wrong

## [1.4.1] - 2017-07-05
### Added
- Experimental anaglyph filter

### Fixed
- Custom filter directory not recognizing relative paths
- CLI nesting self closing tags when injecting SVG into HTML

## [1.4.0] - 2016-11-08
### Added
- Custom filters

### Fixed
- Color filter not merging with source graphic
- Uploading browser dist to npm

## [1.3.1] - 2016-10-26
### Added
- Error if first argument is not an array or a string
- Tests

### Changed
- Compressed SVG filters to one line

### Fixed
- Default SVG filter generating wrong CSS
- Empty SVG element returned if no SVG was generated
- CLI: Saving SVG when no SVG was generated

## [1.3.0] - 2016-10-25
### Added
- Node module
- CLI

### Removed
- jQuery version
- CSS class version

## [1.2.0] - 2016-03-03
### Added
- 6 new vintage SVG filters (vanilla JS version only)

### Changed
- Transition CSS rules are now applied only to Philter elements

### Fixed
- Inconsistent height on SVG filters
- CSS rules being applied to the first selector of the element and breaking on elements with same selectors
- SVG adding to body height
- Filter count increasing faster than data is being returned from the server causing wrong filter ids

## [1.1.2] - 2016-02-28
### Added
- Bower compatibility

## [1.1.1] - 2015-12-15
### Added
- Plugin version in vanilla JavaScript
- Option to remove 'philter' from data attributes to make markup shorter

### Fixed
- False element width or height being set on custom SVG filters

## [1.1.0] - 2015-11-07
### Changed
- Changed classes to data attributes to describe filters

### Removed
- Describing filters with classes The old version still can be found in dist directory


test/test.js
const expect = require('chai').expect
const philter = require('../lib/index')

describe('Module', () => {
  it('should return css and svg as strings', function() {
    philter('<div data-philter-blur="10"></div>', (css, svg)=> {
      expect(css).to.be.a('string')
      expect(svg).to.be.a('string')
    })
  })
  it('should return css and svg as strings with tag: false', function() {
    philter('<div data-blur="10"></div>', {tag: false}, (css, svg)=> {
      expect(css).to.be.a('string')
      expect(svg).to.be.a('string')
    })
  })
  it('should throw an Error when no HTML or files are given', function() {
    expect(philter.bind(philter, (css, svg) => {})).to.throw('Philter: No HTML or files given')
  })
  it('should throw an Error when HTML doesn\'t contain philter data attributes', function() {
    expect(philter.bind(philter, '<div></div>', (css, svg) => {})).to.throw('Philter: No philter data attributes found')
  })
  it('should throw an Error when no callback is given', function() {
    expect(philter.bind(philter, '<div></div>')).to.throw('Philter: Callback must be a function')
  })
  it('should throw an Error when no custom filter directory is given', function() {
    expect(philter.bind(philter, '<div data-philter-custom="test"></div>', {customFilters: ['test']}, (css, svg) => {})).to.throw('Philter: No custom filter directory found')
  })
})

describe('Filters', () => {
  it('blur', function() {
    philter('<div data-philter-blur="10"></div>', (css, svg)=> {
      expect(css).to.equal('[data-philter-blur="10"]{filter:blur(10px);}')
      expect(svg).to.equal('')
    })
  })
  it('grayscale', function() {
    philter('<div data-philter-grayscale="10"></div>', (css, svg)=> {
      expect(css).to.equal('[data-philter-grayscale="10"]{filter:grayscale(10%);}')
      expect(svg).to.equal('')
    })
  })
  it('hue-rotate', function() {
    philter('<div data-philter-hue-rotate="10"></div>', (css, svg)=> {
      expect(css).to.equal('[data-philter-hue-rotate="10"]{filter:hue-rotate(10deg);}')
      expect(svg).to.equal('')
    })
  })
  it('saturate', function() {
    philter('<div data-philter-saturate="10"></div>', (css, svg)=> {
      expect(css).to.equal('[data-philter-saturate="10"]{filter:saturate(10%);}')
      expect(svg).to.equal('')
    })
  })
  it('sepia', function() {
    philter('<div data-philter-sepia="10"></div>', (css, svg)=> {
      expect(css).to.equal('[data-philter-sepia="10"]{filter:sepia(10%);}')
      expect(svg).to.equal('')
    })
  })
  it('contrast', function() {
    philter('<div data-philter-contrast="10"></div>', (css, svg)=> {
      expect(css).to.equal('[data-philter-contrast="10"]{filter:contrast(10%);}')
      expect(svg).to.equal('')
    })
  })
  it('invert', function() {
    philter('<div data-philter-invert="10"></div>', (css, svg)=> {
      expect(css).to.equal('[data-philter-invert="10"]{filter:invert(10%);}')
      expect(svg).to.equal('')
    })
  })
  it('opacity', function() {
    philter('<div data-philter-opacity="10"></div>', (css, svg)=> {
      expect(css).to.equal('[data-philter-opacity="10"]{filter:opacity(10%);}')
      expect(svg).to.equal('')
    })
  })
  it('brightness', function() {
    philter('<div data-philter-brightness="10"></div>', (css, svg)=> {
      expect(css).to.equal('[data-philter-brightness="10"]{filter:brightness(10%);}')
      expect(svg).to.equal('')
    })
  })
  it('drop-shadow', function() {
    philter('<div data-philter-drop-shadow="10 10 10 black"></div>', (css, svg)=> {
      expect(css).to.equal('[data-philter-drop-shadow="10 10 10 black"]{filter:drop-shadow(10px 10px 10px black);}')
      expect(svg).to.equal('')
    })
  })
  it('svg', function() {
    philter('<div data-philter-svg="filter"></div>', (css, svg)=> {
      expect(css).to.equal('[data-philter-svg="filter"]{filter:url(#filter);}')
      expect(svg).to.equal('')
    })
  })
  it('color', function() {
    philter('<div data-philter-color="black 10"></div>', (css, svg)=> {
      expect(css).to.equal('[data-philter-color="black 10"]{filter:url(#color-1);}')
      expect(svg.replace(/\r?\n|\r/g, '')).to.equal('<svg xmlns="http://www.w3.org/2000/svg" version="1.1" width="0" height="0" id="philter-svg"><defs><filter id="color-1" primitiveunits="objectBoundingBox"><feflood x="0" y="0" width="100%" height="100%" flood-opacity="0.1" flood-color="black" in="SourceGraphic" result="overlay"></feflood><feblend in="overlay" in2="SourceGraphic"></feblend></filter></defs></svg>')
    })
  })
  it('vintage', function() {
    philter('<div data-philter-vintage="3"></div>', (css, svg)=> {
      expect(css).to.equal('[data-philter-vintage="3"]{filter:url(#vintage-3);}')
      expect(svg.replace(/\r?\n|\r/g, '')).to.equal('<svg xmlns="http://www.w3.org/2000/svg" version="1.1" width="0" height="0" id="philter-svg"><defs><filter id="vintage-3" primitiveUnits="objectBoundingBox"><feComponentTransfer x="0" y="0" width="100%" height="100%" color-interpolation-filters="sRGB" result="curves"><feFuncR type="table" tableValues="0.196078431372549 0.2196078431372549 0.4352941176470588 0.7843137254901961 1"/><feFuncG type="table" tableValues="0 0.3588235294117647 0.6137254901960784 0.9 1"/><feFuncB type="table" tableValues="0.3196078431372549 0.4333333333333333 0.5156862745098039 0.5980392156862745 0.6882352941176471 0.7745098039215686 0.8529411764705882 0.9823529411764706 1"/><feFuncA type="table" tableValues="1 1"/></feComponentTransfer><feColorMatrix type="saturate" values="0.95" result="desaturate" /><feComponentTransfer x="0" y="0" width="100%" height="100%" color-interpolation-filters="linearRGB" result="exposure"><feFuncR type="gamma" amplitude="1" exponent="1" offset="0.03" /><feFuncG type="gamma" amplitude="1" exponent="1" offset="0.03" /><feFuncB type="gamma" amplitude="1" exponent="1" offset="0.03" /><feFuncA type="gamma" amplitude="1" exponent="1" offset="0" /></feComponentTransfer></filter></defs></svg>')
    })
  })
  it('duotone', function() {
    philter('<div data-philter-duotone="#000000 #ffffff"></div>', (css, svg) => {
      expect(css).to.equal('[data-philter-duotone="#000000 #ffffff"]{filter:url(#duotone-1);}')
      expect(svg.replace(/\r?\n|\r/g, '')).to.equal('<svg xmlns="http://www.w3.org/2000/svg" version="1.1" width="0" height="0" id="philter-svg"><defs><filter id="duotone-1" primitiveunits="objectBoundingBox"><fecolormatrix type="matrix" in="SourceGraphic" result="grayscale" values="1 0 0 0 0 1 0 0 0 0 1 0 0 0 0 0 0 0 1 0"></fecolormatrix><fecomponenttransfer x="0" y="0" width="100%" height="100%" color-interpolation-filters="sRGB" result="gradientMap"><fefuncr type="table" tablevalues="0 1"></fefuncr><fefuncg type="table" tablevalues="0 1"></fefuncg><fefuncb type="table" tablevalues="0 1"></fefuncb><fefunca type="table" tablevalues="1 1"></fefunca></fecomponenttransfer></filter></defs></svg>')
    })
  })
  it('anaglyph', function() {
    philter('<div data-philter-anaglyph="1"></div>', (css, svg)=> {
      expect(css).to.equal('[data-philter-anaglyph="1"]{filter:url(#anaglyph-1);}')
      expect(svg.replace(/\r?\n|\r/g, '')).to.equal('<svg xmlns="http://www.w3.org/2000/svg" version="1.1" width="0" height="0" id="philter-svg"><defs><filter id="anaglyph-1" primitiveunits="objectBoundingBox"><fecomponenttransfer x="0" y="0" width="100%" height="100%" color-interpolation-filters="sRGB" result="red"><fefuncr type="table" tablevalues="0 1"></fefuncr><fefuncg type="table" tablevalues="0 0"></fefuncg><fefuncb type="table" tablevalues="0 0"></fefuncb></fecomponenttransfer><fecomponenttransfer x="0" y="0" width="100%" height="100%" color-interpolation-filters="sRGB" in="SourceGraphic" result="blue"><fefuncr type="table" tablevalues="0 0"></fefuncr><fefuncg type="table" tablevalues="0 1"></fefuncg><fefuncb type="table" tablevalues="0 1"></fefuncb></fecomponenttransfer><feoffset in="red" dx="0.01" result="redOff"></feoffset><feblend in2="blue" in="redOff" x="0.01" y="0" mode="difference"></feblend></filter></defs></svg>')
    })
  })
  it('custom', function() {
    philter('<div data-philter-custom="test"></div>', {customFilterDir: './test/', customFilters: ['test']}, (css, svg) => {
      expect(css).to.equal('[data-philter-custom="test"]{filter:url(#test);}')
      expect(svg.replace(/\r?\n|\r/g, '')).to.equal('<svg xmlns="http://www.w3.org/2000/svg" version="1.1" width="0" height="0" id="philter-svg"><defs><filter id="test" primitiveUnits="objectBoundingBox"></filter></defs></svg>')
    })
  })
})


philter.js
/* Philter v1.0.0 | (c) 2015 Liudas Dzisevicius | MIT License */
(function ($) {

  'use strict';

  function getFilterString (filterStrings, filter, units) {
    if (filter[4]) {
      filterStrings[0] = filterStrings[0] + filter[0] + '(' + filter[1] + units + ' ' + filter[2] + units + ' ' + filter[3] + units + ' #' + filter[4] + ') ';
      if (filter[5] && filter[6] && filter[7] && filter[8] && filter[9]) {
        filterStrings[1] = filterStrings[1] + filter[0] + '(' + filter[6] + units + ' ' + filter[7] + units + ' ' + filter[8] + units + ' #' + filter[9] + ') ';
      } else {
        filterStrings[1] = filterStrings[1] + filter[0] + '(' + filter[1] + units + ' ' + filter[2] + units + ' ' + filter[3] + units + ' #' + filter[4] + ') ';
      }
    } else if (filter[0] == 'svg') {
      filterStrings[0] = filterStrings[0] + 'url(' + units + filter[1] + ') ';
      if (filter[2] && filter[3]) {
        filterStrings[1] = filterStrings[1] + 'url(' + units + filter[3] + ') ';
      } else {
        filterStrings[1] = filterStrings[1] + 'url(' + units + filter[1] + ') ';
      }
    } else {
      filterStrings[0] = filterStrings[0] + filter[0] + '(' + filter[1] + units + ') ';
      if (filter[2] && filter[3]) {
        filterStrings[1] = filterStrings[1] + filter[0] + '(' + filter[3] + units + ') ';
      } else {
        filterStrings[1] = filterStrings[1] + filter[0] + '(' + filter[1] + units + ') ';
      }
    }

    return filterStrings;
  }

  $.philter = function (transitionTime) {

    var filters = [
      'blur',
      'grayscale',
      'hue-rotate',
      'saturate',
      'sepia',
      'contrast',
      'invert',
      'opacity',
      'brightness',
      'drop-shadow',
      'svg'
    ];
    var filtered = [];
    var filterStrings = ['', ''];
    var styleString = '<style>';

    if (!transitionTime) {
      transitionTime = 0.5;
    }

    $('.philter').each(function () {
      var className = $(this).attr('class');
      var classes = className.split(' ');
      className = '';
      for (var i = 0; i < classes.length; i++) {
        for (var j = 0; j < filters.length; j++) {
          if (~classes[i].indexOf(filters[j])) {
            filtered.push(classes[i]);
          }
        }
        classes[i] = '.' + classes[i];
        className = className + classes[i];
      }

      for (var i = 0; i < filtered.length; i++) {
        var filter = filtered[i].split('_');

        switch (filter[0]) {
          case 'blur':
            filterStrings = getFilterString(filterStrings, filter, 'px');
            break;
          case 'hue-rotate':
            filterStrings = getFilterString(filterStrings, filter, 'deg');
            break;
          case 'drop-shadow':
            filterStrings = getFilterString(filterStrings, filter, 'px');
            break;
          case 'svg':
            filterStrings = getFilterString(filterStrings, filter, '#');
            break;
          default:
            filterStrings = getFilterString(filterStrings, filter, '%');
        }
      }

      styleString = styleString + className + '{filter:' + filterStrings[0] + ';-webkit-filter:' + filterStrings[0] + ';}';
      if (filterStrings[0] != filterStrings[1]) {
        styleString = styleString + className + ':hover{filter:' + filterStrings[1] + ';-webkit-filter:' + filterStrings[1] + ';}';
      }

      filterStrings = ['', ''];
      filtered.length = 0;
    });
    styleString = styleString + '.philter{transition:filter ' + transitionTime + 's,-webkit-filter ' + transitionTime + 's;}</style>';
    $('head').append(styleString);
  }
})(jQuery)


js/min/scripts-min.js
$(document).ready(function(){$(".wp1").waypoint(function(){$(".wp1").addClass("animated fadeInUp")},{offset:"75%"}),$(".wp2").waypoint(function(){$(".wp2").addClass("animated fadeInUp")},{offset:"75%"}),$(".wp3").waypoint(function(){$(".wp3").addClass("animated fadeInRight")},{offset:"75%"}),$(".flexslider").flexslider({animation:"slide"}),$(".single_image").fancybox({padding:4}),$('[data-toggle="tooltip"]').tooltip(),$(".nav-toggle").click(function(){$(this).toggleClass("active"),$(".header-nav").toggleClass("open"),event.preventDefault()}),$(".header-nav li a").click(function(){$(".nav-toggle").toggleClass("active"),$(".header-nav").toggleClass("open")}),$(function(){$(window).scroll(function(){var e=$(window).scrollTop();e>=20?($("section.navigation").addClass("fixed"),$("header").css({"border-bottom":"none",padding:"35px 0"}),$("header .member-actions").css({top:"26px"}),$("header .navicon").css({top:"34px"})):($("section.navigation").removeClass("fixed"),$("header").css({"border-bottom":"solid 1px rgba(255, 255, 255, 0.2)",padding:"50px 0"}),$("header .member-actions").css({top:"41px"}),$("header .navicon").css({top:"48px"}))})}),$(function(){$("a[href*=#]:not([href=#])").click(function(){if(location.pathname.replace(/^\//,"")===this.pathname.replace(/^\//,"")&&location.hostname===this.hostname){var e=$(this.hash);if(e=e.length?e:$("[name="+this.hash.slice(1)+"]"),e.length)return $("html,body").animate({scrollTop:e.offset().top},2e3),!1}})})});

lib/index.js
const fs = require('fs')
const path = require('path')
const _ = require('lodash')
const cheerio = require('cheerio')
const Promise = require('bluebird')
const isHtml = require('is-html')
const rgb = require('hex-rgb')

Promise.promisifyAll(fs)

function philter(files, options, cb) {
  if (!_.isString(files) && !_.isArray(files)) {
    throw new Error('Philter: No HTML or files given')
  }
  if (_.isFunction(options)) {
    cb = options
    options = {}
  }
  if (!_.isFunction(cb)) {
    throw new Error('Philter: Callback must be a function')
  }
  options = _.defaults(options, {
    tag: true,
    customFilterDir: '',
    customFilters: []
  })

  if (isHtml(files)) {
    let $ = cheerio.load(files)
    parseElements($, options, cb)
  } else {
    let html = ''
    let promises = []
    _.forEach(files, (file) => {
      promises.push(
        fs.readFileAsync(file, 'utf-8').then((data) => {
          html += data
        }).catch((err) => {
          throw err
        })
      )
    })
    Promise.all(promises).then(() => {
      let $ = cheerio.load(html)
      parseElements($, options, cb)
    })
  }
}

/**
* Parse HTML and return elements with philter attributes
* @param {Object} $ Cheerio object
* @param {Object} filters All possible filters
* @param {Boolean} tag Is 'philter' tag active
*/
function getElements($, filters, tag) {
  let elements = [];
  _.forEach(filters, (unit, filter) => {
    let query;
    if (tag) {
      query = $(`[data-philter-${filter}]`)
    } else {
      query = $(`[data-${filter}]`)
    }
    if (query) {
      _.forEach(query, (element) => {
        if (!_.includes(elements, element)) {
          elements.push(element)
        }
      })
    }
  })

  return elements
}

/**
* Parse elements and pass generated CSS and SVG to the callback
* @param {Object} $ Cheerio object
* @param {Object} options Philter options
* @param {Function} cb Callback
*/
function parseElements($, options, cb) {
  let filters = {
    'blur': 'px',
    'grayscale': '%',
    'hue-rotate': 'deg',
    'saturate': '%',
    'sepia': '%',
    'contrast': '%',
    'invert': '%',
    'opacity': '%',
    'brightness': '%',
    'drop-shadow': (h, v, blur, color) => `${h}px ${v}px ${blur}px ${color}`,
    'svg': (url) => `url(#${url})`,
    'color': (nr) => `url(#color-${nr})`,
    'vintage': (nr) => `url(#vintage-${nr})`,
    'duotone': (nr) => `url(#duotone-${nr})`, 
    'anaglyph': (nr) => `url(#anaglyph-${nr})`,
    'custom': {}
  }
  let filterCount = {
    'color': 0,
    'anaglyph': 0,
    'duotone': 0,
    'vintage-1': 0,
    'vintage-2': 0,
    'vintage-3': 0,
    'vintage-4': 0,
    'vintage-5': 0,
    'vintage-6': 0
  }
  let css = ''
  let svg = ''
  let promises = []
  if (!_.isEmpty(options.customFilters)) {
    if (!options.customFilterDir) {
      throw new Error('Philter: No custom filter directory found')
    }
    _.forEach(options.customFilters, (value) => {
      filters.custom[value] = `url(#${value})`
      filterCount[value] = 0
    })
  }
  let elements = getElements($, filters, options.tag)

  if (_.isEmpty(elements)) {
    throw new Error('Philter: No philter data attributes found')
  }

  _.forEach(elements, (element) => {
    let selector = ''
    let rule = {
      default: '',
      hover: ''
    }
    _.forEach($(element).data(), (value, key) => {
      value = value.toString()
      let values = value.split(' ')
      key = _.replace(_.kebabCase(key), /(philter-)/g, '')
      selector += `[data-${options.tag?'philter-':''}${key}="${value}"]`
      switch (key) {
        case 'color':
          filterCount.color++
          rule.default += filters.color(filterCount.color)
          getColorFilter(values[0], values[1], filterCount.color, promises, (filter) => {
            svg += filter
          })
          if (values[2] && values[3]) {
            filterCount.color++
            rule.hover += filters.color(filterCount.color)
            getColorFilter(values[2], values[3], filterCount.color, promises, (filter) => {
              svg += filter
            })
          }
          break
        case 'anaglyph':
          filterCount.anaglyph++
          rule.default += filters.anaglyph(filterCount.anaglyph)
          getAnaglyphFilter(values[0], filterCount.anaglyph, promises, (filter) => {
            svg += filter
          })
          if (values[1]) {
            filterCount.anaglyph++
            rule.hover += filters.anaglyph(filterCount.anaglyph)
            getAnaglyphFilter(values[1], filterCount.anaglyph, promises, (filter) => {
              svg += filter
            })
          }
          break
        case 'duotone':
          filterCount.duotone++
          rule.default += filters.duotone(filterCount.duotone)
          getDuotoneFilter(values[0], values[1], filterCount.duotone, promises, (filter) => {
            svg += filter
          })
          if (values[2] && values[3]) {
            filterCount.duotone++
            rule.hover += filters.duotone(filterCount.duotone)
            getDuotoneFilter(values[2], values[3], filterCount.duotone, promises, (filter) => {
              svg += filter
            })
          }
          break
        case 'vintage':
          if (!_.has(filterCount, `vintage-${values[0]}`)) {
            throw new Error(`Philter: No such filter: vintage-${values[0]}`);
          }
          filterCount[`vintage-${values[0]}`]++;
          rule.default += filters.vintage(values[0])
          if (filterCount[`vintage-${values[0]}`] === 1) {
            getVintageFilter(values[0], promises, (filter) => {
              svg += filter
            })
          }
          if (values[1]) {
            filterCount[`vintage-${values[1]}`]++;
            rule.hover += filters.vintage(values[1])
            if (filterCount[`vintage-${values[1]}`] === 1) {
              getVintageFilter(values[1], promises, (filter) => {
                svg += filter
              })
            }
          }
          break
        case 'drop-shadow':
          rule.default += `${key}(${filters[key](values[0], values[1], values[2], values[3])})`
          if (values[4] && values[5] && values[6] && values[7]) {
            rule.hover += `${key}(${filters[key](values[4], values[5], values[6], values[7])})`
          }
          break
        case 'svg':
          rule.default += filters.svg(values[0])
          if (values[1]) {
            rule.hover += filters.svg(values[1])
          }
          break
        case 'custom':
          filterCount[values[0]]++
          rule.default += filters.custom[values[0]]
          if (filterCount[values[0]] === 1) {
            getCustomFilter(path.join(options.customFilterDir, values[0]), promises, (filter) => {
              svg += filter
            })
          }
          if (values[1]) {
            filterCount[values[1]]
            rule.hover += custom.filters[values[1]]
            if (filterCount[values[1]] === 1) {
              getCustomFilter(path.join(options.customFilterDir, values[1]), promises, (filter) => {
                svg += filter
              })
            }
          }
          break
        default:
          // setup default rule
          rule.default += `${key}(${values[0]+filters[key]})`

          // setup hover rule
          if (values[1]) {
            rule.hover += `${key}(${values[1]+filters[key]})`
          }
      }
    })
    css += `${selector}{filter:${rule.default};}`
    if (rule.hover) {
      css += `${selector}:hover{filter:${rule.hover};}`
    }

    selector = ''
    rule.default = ''
    rule.hover = ''
  })
  Promise.all(promises).then(() =>   {
    let svgWrapper = ''
    if (svg) {
      svgWrapper = `<svg xmlns="http://www.w3.org/2000/svg" version="1.1" width="0" height="0" id="philter-svg"><defs>${svg}</defs></svg>`
    } else {
      svgWrapper = ''
    }

    cb(css, svgWrapper)
  })
}

/**
* Parse color filter and pass it to the callback
* @param {String} color Filter color
* @param {String} opacity Filter opacity
* @param {Integer} id Filter ID
* @param {Array} promises List of readFileAsync promises
* @param {Function} cb Callback
*/
function getColorFilter(color, opacity, id, promises, cb) {
  let filter = ''
  promises.push(
    fs.readFileAsync(`${__dirname}/svg/color.svg`, 'utf-8').then((data) => {
      let $ = cheerio.load(data, {recognizeSelfClosing: true});
      $('filter').attr('id', 'color-' + id)
      $('feflood').attr('flood-color', color).attr('flood-opacity', opacity/100)
      filter = $.html()
      cb(filter)
    }).catch((err) => {
      throw err
    })
  )
}

/**
* Parse anaglyph filter and pass it to the callback
* @param {String} offset Filter offset
* @param {Integer} id Filter ID
* @param {Array} promises List of readFileAsync promises
* @param {Function} cb Callback
*/
function getAnaglyphFilter(offset, id, promises, cb) {
  let filter = ''
  promises.push(
    fs.readFileAsync(`${__dirname}/svg/anaglyph.svg`, 'utf-8').then((data) => {
      let $ = cheerio.load(data, {recognizeSelfClosing: true});
      $('filter').attr('id', 'anaglyph-' + id)
      $('feoffset').attr('dx', offset/100)
      $('feblend').attr('x', offset/100)
      filter = $.html()
      cb(filter)
    }).catch((err) => {
      throw err
    })
  )
}

/**
* Parse duotone filter and pass it to the callback
* @param {String} color1 First filter color
* @param {String} color2 Second filter color
* @param {Integer} id Filter ID
* @param {Array} promises List of readFileAsync promises
* @param {Function} cb Callback
*/
function getDuotoneFilter(color1, color2, id, promises, cb) {
  let filter = ''
  color1 = rgb(color1);
  color2 = rgb(color2);
  promises.push(
    fs.readFileAsync(`${__dirname}/svg/duotone.svg`, 'utf-8').then((data) => {
      let $ = cheerio.load(data, {recognizeSelfClosing: true});
      $('filter').attr('id', 'duotone-' + id)
      $('fefuncr').attr('tablevalues', color1.red/255+' '+color2.red/255)
      $('fefuncg').attr('tablevalues', color1.green/255+' '+color2.green/255)
      $('fefuncb').attr('tablevalues', color1.blue/255+' '+color2.blue/255)
      filter = $.html()
      cb(filter)
    }).catch((err) => {
      throw err
    })
  )
}

/**
* Read vintage filter and pass it to the callback
* @param {Integer} id Filter ID
* @param {Array} promises List of readFileAsync promises
* @param {Function} cb Callback
*/
function getVintageFilter(id, promises, cb) {
  let filter = ''
  promises.push(
    fs.readFileAsync(`${__dirname}/svg/vintage-${id}.svg`, 'utf-8').then((data) => {
      cb(data)
    }).catch((err) => {
      throw err
    })
  )
}

/**
* Load custom SVG filter
* @param {String} file SVG file directory
* @param {Array} promises List of readFileAsync promises
* @param {Function} cb Callback
*/
function getCustomFilter(dir, promises, cb) {
  promises.push(
    fs.readFileAsync(dir + '.svg', 'utf-8').then((data) => {
      cb(data)
    }).catch((err) => {
      throw err
    })
  )
}

module.exports = philter
